.travis.yml

```diff
@@ -14,24 +14,147 @@ git:
 
 sudo: false
 
+cache:
+  ccache: true
+  apt: true
+  directories:
+    - "${HOME}/mpich"
+    - "${HOME}/.ccache"
+
 env:
   global:
     - INSTALLDIR="$HOME/taucmdr-test"
-  matrix:
-    - OS='Ubuntu Precise' # Add more python versions to this array to expand test matrix
+    - CC="ccache gcc${SUF:-}" CXX="ccache g++${SUF:-}" FC="ccache gfortran${SUF:-}" F77="ccache gfortran${SUF:-}"
+
+matrix:
+  include:
+    - &baseconf # define an anchor and use the <<: YAML merge operator to avoid repitition
+      python: 2.7
+      cache:
+        ccache: true
+        apt: true
+        directories:
+          - "${HOME}/mpich"
+          - "${HOME}/.ccache"
+      addons:
+        apt:
+          packages: ['gcc-4.4', 'g++-4.4', 'gfortran-4.4', 'ccache']
+      env: GCC=-4.4 MPI=none
+    -
+      <<: *baseconf
+      env: GCC=-4.4 MPI=MPICH
+    - &fourfive
+      <<: *baseconf
+      addons:
+        apt:
+          packages: ['gcc-4.5', 'g++-4.5', 'gfortran-4.5', 'ccache']
+      env: GCC=-4.5 MPI=none
+    -
+      <<: *fourfive
+      env: GCC=-4.5 MPI=MPICH
+    - &foursix
+      <<: *baseconf
+      addons:
+        apt:
+          packages: ['gcc-4.6', 'g++-4.6', 'gfortran-4.6', 'ccache']
+      env: GCC=-4.6 MPI=none
+    -
+      <<: *foursix
+      env: GCC=-4.6 MPI=MPICH
+    - &fourseven
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.7', 'g++-4.7', 'gfortran-4.7', 'ccache']
+      env: GCC=-4.7 MPI=none
+    -
+      <<: *fourseven
+      env: GCC=-4.7 MPI=MPICH
+    - &foureight
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.8', 'g++-4.8', 'gfortran-4.8', 'ccache']
+      env: GCC=-4.8 MPI=none
+    -
+      <<: *foureight
+      env: GCC=-4.8 MPI=MPICH
+    - &fournine
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.9', 'g++-4.9', 'gfortran-4.9', 'ccache']
+      env: GCC=-4.9 MPI=none
+    -
+      <<: *fournine
+      env: GCC=-4.9 MPI=MPICH
+    - &five
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-5', 'g++-5', 'gfortran-5', 'ccache']
+      env: GCC=-5 MPI=none
+
+    -
+      <<: *five
+      env: GCC=-5 MPI=MPICH
+    - &six
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-6', 'g++-6', 'gfortran-6', 'ccache']
+      env: GCC=-6 MPI=none
+    -
+      <<: *six
+      env: GCC=-6 MPI=MPICH
```

Comment:
Unfortunately all the configs need to be enumerated via `matrix.include`. However, YAML anchors, references and merges help make this a little bit less painful. Just be careful that it's valid YAML and a valid .travis.yml file. `gem install travis ; travis lint` is a good place to start before pushing commits.